linked_list/linked_list.py

```diff
@@ -9,71 +9,130 @@ def __init__(self, value, next_node = None):
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter = 0
+        current = self.head
+
+        while current != None:
+            current = current.next
+            counter += 1
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current_index = 0
+        current = self.head
+        while self.head != None:
+            if current_index == index:
+                return current.value
+            else:
+                current = current.next
+                current_index += 1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        current = self.head
+        while current.next != None:
+            current = current.next
+        return current.value
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        last_node = Node(value)
+
+        if self.head == None:
+            self.head = last_node
+            return self.head.value
+        
+        current = self.head
+        while current.next != None:
+            current = current.next
+        current.next = last_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        max_value = self.head.value
+        current = self.head.next
+
+        while current != None:
+            if current.value > max_value:
+                max_value = current.value
+            current = current.next
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):

```

Comment:
👍 , I do want to point out that this implementation is deleting all instances of `value` in the list.  Not wrong, just an implementation choice.